### PR TITLE
Lwm2m block transfer

### PR DESCRIPTION
**Pull request by LukasWoodtli**
When initializing a lwm2m message it's possible to request a big buffer for serializing the message body. Before sending a message with the big buffer in use, the body is split into blocks.

This approach was chosen due to several reasons.:

- The implementation is mostly decoupled from the existing code. If block transfer is not needed, the corresponding code is excluded by preprocessor defines (conditional compilation).
- No need to adjust any code that generates message content (headers, options, body). The existing code just writes to the message buffer as before. But the buffer can be bigger (if requested) than the actual lwm2m packet size.

There is still some (or a lot) room for improvement:

How to decide if a big buffer is needed for serializing the message? Currently, it's the responsibility of the developer. But there could be a heuristic or another algorithm for deciding if a big buffer should be requested. It could also be set by one or multiple configurations.

There is some copying of buffers involved. This could probably be improved in some cases.

This implementation is just for lwm2m even if block-transfer is a CoAP feature.

### CODE REVIEW

**Comment by LukasWoodtli** [comment]
I'm working on some small improvements and fixes. So please be aware that some details will change. But the basic idea and the overall code structure will still be the same.

**Comment by LukasWoodtli** [comment]
This PR (draft) is ready for review.
Please give me some feedback. Thanks!

**Comment by SeppoTakalo** [comment]
@LukasWoodtli I'm not able to find out the reason, so why are we not using the big buffer for all the messages?
Why only some (registration, Notify)?

I cannot seem to be able to trigger this feature just using the `native_posix` build.

I wonder if we just always render the packet into a big buffer, the split it into multiple BLOCK* CoAPs.

**Comment by LukasWoodtli** [comment]
> I'm not able to find out the reason, so why are we not using the big buffer for all the messages?

The choice where to request a big buffer is somewhat arbitrary. I had a look at the code to see where a big message body could potentially be created.
How to decide where a big buffer is requested is definitely a topic for discussion and there is room for improvement.

**Comment by LukasWoodtli** [comment]
> I cannot seem to be able to trigger this feature just using the `native_posix` build.

To trigger this feature you need to set `CONFIG_LWM2M_COAP_BLOCK_SIZE` to a value that is a multiple of `CONFIG_LWM2M_COAP_MAX_MSG_SIZE`.

**Comment by LukasWoodtli** [comment]
> I wonder if we just always render the packet into a big buffer, the split it into multiple BLOCK* CoAPs.

Yes this would be a possibility. But it would possibly require a bigger buffer pool. On the other hand it would reduce the complexity of the code a bit.

**Comment by SeppoTakalo** [comment]
In my mind, if we should support this, it should be supported for all packets, not just some.

I tried to trigger this feature, but it was not supported for READ requests, and I believe  I also failed to trigger this on SEND command. So it is a bit hard to test.
I think we should aim for less complexity and create packets always on big buffer, instead of just some selected messages.

We also need to take a look that how many ongoing requests we even could have.. Is it going to block a lot if there is just one big buffer to be used.

**Comment by LukasWoodtli** [comment]
> In my mind, if we should support this, it should be supported for all packets, not just some.
> 
> I tried to trigger this feature, but it was not supported for READ requests, and I believe I also failed to trigger this on SEND command. So it is a bit hard to test. I think we should aim for less complexity and create packets always on big buffer, instead of just some selected messages.
> 
> We also need to take a look that how many ongoing requests we even could have.. Is it going to block a lot if there is just one big buffer to be used.

I have added some complexity which could be considered as premature optimization. I’m happy to simplify this part of the code.

Regarding number of big buffers: if we activate this feature for all messages we will probably need more big buffers in the pool.

I’ll have a look at it asap. But it won’t be this week anymore.

File: subsys/net/lib/coap/coap.c
Comment (rlubos) [inline subsys/net/lib/coap/coap.c:980]: I don't actually think this is correct - both Block1 and Block2 options can be either part of request or response. The difference on the use of those options is whether you request data (GET) or proactively send (PUT/POST).

File: subsys/net/lib/lwm2m/Kconfig
Comment (rlubos) [inline subsys/net/lib/lwm2m/Kconfig:114]: I don't think I understand why do we need such a requirement. And I don't think it should actually be used the way it is, i.e. to determine the "large buffer" size. The option description is pretty clear - "CoAP message size used by LWM2M."

File: subsys/net/lib/lwm2m/lwm2m_object.h
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_object.h:139]: I think this should simply be a Kconfig option. Plus probably another one to determine the actual size of the buffer for large message serialization.

**Comment by rlubos** [comment]
> I wonder if we just always render the packet into a big buffer, the split it into multiple BLOCK* CoAPs.

I second such approach. I think this is the most feasible way to determine the actual payload size w/o rewriting content encoders heavily. We could have a preallocated buffer, mutex protected, used only for encoding. Then, if it the message size would exceed the predefined packet size, allocate a dedicated buffer for the block transfer, and send block by block.

I don't think we should split into multiple coap messages prematurely, as this kind of blows the idea of block size negotiation.

File: subsys/net/lib/lwm2m/Kconfig
Comment (LukasWoodtli) [reply subsys/net/lib/lwm2m/Kconfig:114]: This is not a strict requirement. I can remove that.

File: subsys/net/lib/lwm2m/lwm2m_object.h
Comment (LukasWoodtli) [reply subsys/net/lib/lwm2m/lwm2m_object.h:139]: I'm happy to add a Kconfig for `LWM2M_COAP_BLOCK_TRANSFER_ENABLED`. Maybe anyone has a better name for the config.

But I'm a bit confused about the config for the big buffer. Then we would have tree different sizes for buffers:
1. block size
2. message size
3. big buffer size for serialization

Probably I don't understand the meaning of the message size.

**Comment by LukasWoodtli** [comment]
> > I wonder if we just always render the packet into a big buffer, the split it into multiple BLOCK* CoAPs.
> 
> I second such approach. I think this is the most feasible way to determine the actual payload size w/o rewriting content encoders heavily. We could have a preallocated buffer, mutex protected, used only for encoding. Then, if it the message size would exceed the predefined packet size, allocate a dedicated buffer for the block transfer, and send block by block.
> 
> I don't think we should split into multiple coap messages prematurely, as this kind of blows the idea of block size negotiation.

My PR is not handling block size negotiation yet. But it would be quite easy to introduce. Currently, the content of a new block to send is determined by the 'fixed' block size and the block number in the response with the CONTINUE code. But with some bookkeeping, it would be possible to handle the case of block size negotiation.
I suggest adding that in a separate PR.

Otherwise, the approach is as described. Splitting blocks out of the buffer as needed.

File: subsys/net/lib/coap/coap.c
Comment (rlubos) [reply subsys/net/lib/coap/coap.c:980]: Well, the function name doesn't suggest that. IMO if we provide such a generic function, it should cover all cases.

File: subsys/net/lib/lwm2m/lwm2m_object.h
Comment (rlubos) [reply subsys/net/lib/lwm2m/lwm2m_object.h]: Message size represents the maximum CoAP payload size the message can carry. In most cases, it should be lower than network MTU (minus CoAP/UDP/IP header sizes) so that the message can fit a single UDP datagram. In case IP fragmentation is supported though, one could try to use even larger message sizes, however I don't think we've ever tested such case.

So to sum up:
* `LWM2M_COAP_BLOCK_SIZE` represents the CoAP block size size used by LwM2M, this is pretty straightforward
* `LWM2M_COAP_MAX_MSG_SIZE` represents the maximum payload a **single** CoAP packet can carry. This should be no lower than `LWM2M_COAP_BLOCK_SIZE`, and block transfer should only be involved if the desired payload size to transmit would exceed `LWM2M_COAP_MAX_MSG_SIZE`
* Plus, as you've noticed, we'd need a third config for the serialization buffer. This should be higher than `LWM2M_COAP_MAX_MSG_SIZE`, as otherwise it'd make little sense (whole serialization buffer would fit a single CoAP message, so there's no really need for block transfer).

File: subsys/net/lib/coap/coap.c
Comment (LukasWoodtli) [reply subsys/net/lib/coap/coap.c:980]: I would prefer to name the function properly (i.e. `coap_append_descriptive_block_option`) instead of implementing a generic function that we don't need yet (YAGNI).

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_message_handling.c:2534]: If you don't reset the fragment message, who releases the reply/pending structs then? A new one is allocated for the next fragment unless I miss something.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c:494]: This part seems not needed, `-EINVAL` is already set in `ret`.

Comment (rlubos) [reply subsys/net/lib/lwm2m/lwm2m_rd_client.c:494]: Also I wonder if this could be hidden somehow, so that the callback is only called when the callback is complete and the final response code is received.

**Review by rlubos — COMMENTED** [review]
Sorry for late review. Looks like a step in the right direction (at least to me) but I still have some concerns wrt memory management or CON handling (pendings).

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c
Comment (LukasWoodtli) [reply subsys/net/lib/lwm2m/lwm2m_message_handling.c:243]: To my understanding, some implementations of CoAP (and LWM2M) use the token to decide which blocks belong together to one message body. That's why we reuse the token here.
Zephyr also uses the token on the receiving side to find the right block context for a received block.

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c
Comment (LukasWoodtli) [reply subsys/net/lib/lwm2m/lwm2m_message_handling.c:2534]: In `build_msg_block_for_send(..)` we call `lwm2m_reset_message(msg, false)`. There the `pending`and the `reply` pointers of the message are cleared.

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c
Comment (LukasWoodtli) [reply subsys/net/lib/lwm2m/lwm2m_message_handling.c:243]: And yes, I missed to reuse the token from the response.

File: tests/net/lib/lwm2m/block_transfer/src/main.c
Comment (rlubos) [inline tests/net/lib/lwm2m/block_transfer/src/main.c:2]: It'd be nice to split tests into a separate commit, for clarity.

**Review by rlubos — COMMENTED** [review]
Thanks for the update @LukasWoodtli. I think we're on par with the actual design, but it'd be good to get a second review on the actual changes.

**Comment by rlubos** [comment]
@SeppoTakalo Please have a look at convenience.

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c
Comment (LukasWoodtli) [reply subsys/net/lib/lwm2m/lwm2m_message_handling.c:368]: True. This makes sense!

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c
Comment (LukasWoodtli) [reply subsys/net/lib/lwm2m/lwm2m_message_handling.c:354]: Yes this function is called between writing the message (to the big encode buffer) and adding it to the message queue for sending. It decides if the message needs block-transfer. If so it creates the first block and keeps the encode buffer for the following blocks. In the case where block-transfer is not needed it just copies the data from the encode buffer and releases it.
This function is then not called anymore. Handling of following blocks is is done with `build_msg_block_for_send()`.

File: subsys/net/lib/coap/coap.c
Comment (LukasWoodtli) [reply subsys/net/lib/coap/coap.c:977]: In some cases we could remove a call to `is_request()` in an `if`. But it would require a refactoring. This is probably not worth the effort.

**Review by rlubos — APPROVED** [review]
No more comments from my side, thanks.

**Review by SeppoTakalo — APPROVED** [review]

**Comment by lodup29** [comment]
I experimented with block transfers as introduced here. A few observations:

- CoAP response code 2.31 (COAP_RESPONSE_CODE_CONTINUE) is not handled everywhere. [do_send_reply_cb](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/lwm2m/lwm2m_message_handling.c#L3197) is called at one point with this code. As a result, the callback registered by lwm2m_send_cb is called with a failure code but on the line, everything looks good:
send_ok
- Reading the resource fails. The client keeps responding with block # 0 even if block # 1 is requested:
read_fail
- I think I would make sense to be able to build the payload block by block on the fly with block aware callbacks for read and send operations [as is it the case for the write operation](https://github.com/zephyrproject-rtos/zephyr/blob/d01780fc9403166d250ded3d02372a7e1a9db4fd/include/zephyr/net/lwm2m.h#L290).

**Comment by SeppoTakalo** [comment]
> I think it would make sense to be able to build the payload block by block on the fly with block aware callbacks for read and send operations [as is it the case for the write operation](https://github.com/zephyrproject-rtos/zephyr/blob/d01780fc9403166d250ded3d02372a7e1a9db4fd/include/zephyr/net/lwm2m.h#L290).

This is not entirely true. The block-wise write only works for opaque binary data. So only when you write a binary blob into one resource.

It does not work when the data is encoded as our encoders expects to get the full payload and parse it completely. For example CBOR parsers expects this. Some LwM2M specific parsers might be able to be tweaked to support partial data, but none are currently able to. For example getting LwM2M composite write to work with CoAP block-wise would require this. Any write that has more than one resource requires encoded payload.

**Comment by LukasWoodtli** [comment]
> I experimented with block transfers as introduced here. A few observations:
> 
>     * CoAP response code 2.31 (COAP_RESPONSE_CODE_CONTINUE) is not handled everywhere. [do_send_reply_cb](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/lwm2m/lwm2m_message_handling.c#L3197) is called after the first block with this code. As a result, the callback registered by lwm2m_send_cb is called with a failure code but on the line, everything looks good:
>       send_ok
> 
>     * Reading the resource fails. The client keeps responding with block # 0 even if block # 1 is requested:
>       read_fail
> 
>     * I think it would make sense to be able to build the payload block by block on the fly with block aware callbacks for read and send operations [as is it the case for the write operation](https://github.com/zephyrproject-rtos/zephyr/blob/d01780fc9403166d250ded3d02372a7e1a9db4fd/include/zephyr/net/lwm2m.h#L290).

Could you please check if  #60764 solves your problem?

**Pull request merged**
